### DISCUSSION TITLE
Codebase exploration help

### DISCUSSION THREAD

**Discussion by chayan-blip** [ben-manes/caffeine#1343] [General]
Hi Ben,

This is a beginner question, but my understanding of Object-Oriented Programming is that just like a car there will be a "top level object" and a "top level test suite" for example which checks when pedals and steering is applied to the car it moves in the "correct direction". Similarly I was hoping to find the "top level object and its corresponding test suite" and go to the subcomponents down from there. Is this a correct ask? Please help me in exploring the codebase. I am trying to understand the purpose of each file and any help regarding the same will be highly helpful. 

Thanks
Chayan

**Comment by ben-manes** [comment]
A data structure, and especially a concurrent one, will tend to disfavor OO for performance reasons. Thus, the implementation will follow different best practices than application code.

The primary classes of interest are,
- [BoundedLocalCache](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java): A bounded map with eviction policies.
- [Node](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Node.java): The cache entry, which has code generated subclasses (see [javapoet](https://github.com/ben-manes/caffeine/tree/master/caffeine/src/javaPoet)).
- [BoundedBuffer](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedBuffer.java) and [StripedBuffer](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/StripedBuffer.java): A lossy ring buffer to sample recent entry reads to replay onto the policy
- [MpscGrowableArrayQueue](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java): A write buffer to batch entry writes to replay onto the policy
- [FrequencySketch](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java): A compact histogram to estimate an entry's frequency.
- [TimerWheel](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/TimerWheel.java): A time-based priority queue for expiration
- [Pacer](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Pacer.java): A scheduler for expiration removal to that ensures a slight delay to allow for better batching behavior.

These are discussed at a high-level in the [design document](https://github.com/ben-manes/caffeine/wiki/Design). The remaining classes are primary scaffolding and interfaces to the external API. 

The tests contain our [suite](https://github.com/ben-manes/caffeine/tree/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache) as well as importing other's for cross checking (apache, eclipse, google, jdk, lincheck). To increase coverage, our tests are parameterized to run with every valid configuration using [CacheSpec](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/testing/CacheSpec.java) to instruct the [CacheProvider](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/testing/CacheProvider.java). The test methods are kept short by extracting the assertions checks into Truth Subjects, e.g. [CacheSubject](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/testing/CacheSubject.java). The test coverage is very broad because the implementation complexity is too high to keep every detail in my head.
